rig/machine_control/machine_controller.py

```diff
@@ -1739,34 +1807,169 @@ def get_machine(self, x=0, y=0, default_num_cores=18):
+        fragmentation this constraint will make a conservative estimate based
+        on the largest free block of SDRAM/SRAM in the heap.
+
+        Parameters
+        ----------
+        machine : :py:class:`~rig.machine.Machine`
+            The :py:class:`~rig.machine.Machine` model of the current SpiNNaker
+            machine (e.g. from :py:meth:`get_machine`).
+        core_resource : resource (Default: :py:data:`~rig.machine.Cores`)
+            The resource identifier used for cores. If using
+            :py:meth:`get_machine`, the default correct.
+        sdram_resource : resource (Default: :py:data:`~rig.machine.SDRAM`)
+            The resource identifier used for SDRAM. If using
+            :py:meth:`get_machine`, the default correct.
+        sram_resource : resource (Default: :py:data:`~rig.machine.SRAM`)
+            The resource identifier used for SRAM. If using
+            :py:meth:`get_machine`, the default correct.
+        _all_chips_info : None
+            For internal use only. Users should leave this as None.
+
+        Returns
+        -------
+        [\
+        :py:class:`rig.place_and_route.constraints.ReserveResourceConstraint`,\
+        ...]
+            A set of place-and-route constraints which reserves non-idle cores
+            and SDRAM. The resource types given in the ``core_resource`` and
+            ``sdram_resource`` arguments will be reserved accordingly.
+        """
+        if _all_chips_info is not None:
+            chips_info = _all_chips_info[2]
+        else:
+            chips_info = {}
+
+        constraints = []
+
+        for x, y in machine:
+            chip_info = chips_info.get((x, y))
+            if chip_info is None:
+                chip_info = self.get_chip_info(x, y)
+
+            # Construct a minimal set of ReserveResourceConstraints which
+            # reserve the cores which are not in the idle state.
+            cur_reservation = None
+            for state, core in zip(chip_info.core_states,
+                                   range(machine[(x, y)][core_resource])):
+                if state == consts.AppState.idle:
+                    if cur_reservation is not None:
+                        constraints.append(cur_reservation)
+                    cur_reservation = None
+                else:
+                    if cur_reservation is None:
+                        cur_reservation = ReserveResourceConstraint(
+                            core_resource,
+                            slice(core, core + 1),
+                            (x, y))
+                    else:
+                        cur_reservation.reservation = slice(
+                            cur_reservation.reservation.start,
+                            core + 1)
+            if cur_reservation is not None:
+                constraints.append(cur_reservation)
+
+            # Reserve enough SDRAM such that the amount remaining will be the
+            # size of the largest free block in the heap.
```

Comment:
We're relying on there being little fragmentation. Is that wise?

[NB I think it is probably fine, assuming that SARK behaves sanely when freeing and that most applications can be expected to make few, large, requests for memory]